imi/src/main/java/org/motechproject/nms/imi/service/impl/CdrFileServiceImpl.java

```diff
@@ -42,167 +59,426 @@
-            records.put(cdr.getRequestId().toString(), record);
+    private int getMaxErrorCount() {
+        try {
+            return Integer.parseInt(settingsFacade.getProperty(MAX_CDR_ERROR_COUNT));
+        } catch (NumberFormatException e) {
+            return MAX_CDR_ERROR_COUNT_DEFAULT;
         }
     }
 
 
-    // Create an in-memory set of summary records from all detail records
-    private ParseResults parseDetailRecords(File file, FileInfo fileInfo) {
-        String thisChecksum = "";
-        Map<String, CallSummaryRecordDto> records = new HashMap<>();
+    private void sendProcessSummaryRecordEvent(CallSummaryRecordDto record) {
+        Map<String, Object> eventParams = new HashMap<>();
+        eventParams.put(CSR_PARAM_KEY, record);
+        MotechEvent motechEvent = new MotechEvent(PROCESS_SUMMARY_RECORD_SUBJECT, eventParams);
+        eventRelay.sendEventMessage(motechEvent);
+    }
+
+
+    /**
+     *
+     * @param file          the actual file to process
+     * @param fileInfo      file information provided about the file (ie: expected checksum & recordCount)
+     * @param action        PASS1: verify checksum, verify record count, verify CSV valid
+     *                      PASS2: PASS1 + verify individual records point to existing entities, check sort order
+     *                      PASS3: PASS1 + aggregate same call CDRS into CSR and send for distributed processing
+     * @return
+     */
+    @Override // NO CHECKSTYLE Cyclomatic Complexity
+    @SuppressWarnings("PMD.NcssMethodCount")
+    public List<String> iterateDetailFile(File file, FileInfo fileInfo, Action action) {
+        int maxErrorCount = getMaxErrorCount();
+        int errorCount = 0;
         List<String> errors = new ArrayList<>();
+        Set<String> requestIds = new HashSet<>();
         int lineNumber = 1;
+        String thisChecksum = "";
+        // Might end with .sorted (PASS2, PASS3) or not (PASS1)
+        String fileName = file.getName();
 
-        try (
-                FileInputStream fis = new FileInputStream(file);
-                InputStreamReader isr = new InputStreamReader(fis);
-                BufferedReader reader = new BufferedReader(isr)
-            ) {
-
+        try (FileInputStream fis = new FileInputStream(file);
+             InputStreamReader isr = new InputStreamReader(fis);
+             BufferedReader reader = new BufferedReader(isr)) {
 
-            MessageDigest md = MessageDigest.getInstance("MD5");
+            MessageDigest md = null;
             @SuppressWarnings("PMD.UnusedLocalVariable")
-            DigestInputStream dis = new DigestInputStream(fis, md);
+            DigestInputStream dis = null;
+            if (action == Action.PASS1) {
+                md = MessageDigest.getInstance("MD5");
+                dis = new DigestInputStream(fis, md);
+            }
 
             String line;
+            String currentRequestId = "";
+            CallSummaryRecordDto csr = null;
```

Comment:
I'm having trouble keeping up with all the different things that happen in this method. If I understand it correctly, the commonality between the passes is the iteration of the file, correct? 

I feel like the while loop should should be switch-cased on the pass or break it down into 3 different methods. I'm starting to side with the pmd on this one :cry: